zookeeper-server/src/main/java/org/apache/zookeeper/server/backup/RestoreFromBackupTool.java

```diff
@@ -0,0 +1,375 @@
+   * @param zxidToRestore the zxid upto which to restore, or Long.MAX to restore to the latest
+   *                      available transactionally consistent zxid.
+   * @param dryRun whether this is a dryrun in which case no files are actually copied
+   * @param restoreTempDir a temporary, local (not remote) directory to stage the backup files from
+   *                       remote backup storage for the processing stage of restoration. Note that
+   *                       this directory and the files in it will be removed after restoration.
+   */
+  RestoreFromBackupTool(BackupStorageProvider storageProvider, FileTxnSnapLog snapLog,
+      long zxidToRestore, boolean dryRun, File restoreTempDir) {
+
+    filesToCopy = new ArrayList<>();
+    snapNeededIndex = -1;
+    mostRecentLogNeededIndex = -1;
+    oldestLogNeededIndex = -1;
+
+    this.storage = storageProvider;
+    this.zxidToRestore = zxidToRestore;
+    this.snapLog = snapLog;
+    this.dryRun = dryRun;
+    this.restoreTempDir = restoreTempDir;
+  }
+
+  /**
+   * Attempts to perform a restore.
+   */
+  public void run() throws IOException {
+    try {
+      if (!findFilesToRestore()) {
+        throw new IllegalArgumentException("Failed to find valid snapshot and logs to restore.");
+      }
+
+      if (snapLog == null || restoreTempDir == null || storage == null) {
+        throw new BackupException(
+            "The FileTxnSnapLog, RestoreTempDir and BackupStorageProvider cannot be null.");
+      }
+
+      File dataDir = snapLog.getDataDir();
+      File snapDir = snapLog.getSnapDir();
+      if (!dataDir.exists() && !dataDir.mkdirs()) {
+        throw new BackupException("Failed to create a data directory at path: " + dataDir.getPath()
+            + " to store restored txn logs.");
+      }
+      if (!snapDir.exists() && !snapDir.mkdirs()) {
+        throw new BackupException("Failed to create a snap directory at path: " + snapDir.getPath()
+            + " to store restored snapshot files.");
+      }
+
+      if (!dryRun) {
+        if (!restoreTempDir.exists() && !restoreTempDir.mkdirs()) {
+          throw new BackupException(
+              "Failed to create a temporary directory at path: " + restoreTempDir.getPath()
+                  + " to store copied backup files.");
+        }
+
+        // This step will create a "version-2" directory inside restoreTempDir,
+        // all the selected backup files will be copied to version-2 directory
+        FileTxnSnapLog restoreTempSnapLog =
+            new FileTxnSnapLog(this.restoreTempDir, this.restoreTempDir);
+
+        copyBackupFilesToLocalTempDir(restoreTempSnapLog);
+        processCopiedBackupFiles(restoreTempSnapLog, zxidToRestore);
+        copyProcessedRestoredFilesToDestination(restoreTempSnapLog);
+      }
+    } finally {
+      if (restoreTempDir != null && restoreTempDir.exists()) {
+        BackupStorageUtil.deleteDirectoryRecursively(restoreTempDir);
```

Comment:
Nit: let's add a quick comment here why we need to delete recursively - this is because "version-2" gets created with FileTxnSnapLog.